File: backend/src/schemas/organization.py
from typing import Any

"""
组织架构相关数据验证模式

注意：组织类型和状态字段已使用 str 类型，业务规则验证由 Service 层处理
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator
from sqlalchemy import inspect as sa_inspect
from sqlalchemy.orm import attributes as orm_attributes

from ..models.organization import RepresentedPartyPerspective


def _validate_represented_party_pair(
    represented_party_id: str | None,
    represented_party_perspective: RepresentedPartyPerspective | None,
) -> None:
    if (represented_party_id is None) != (represented_party_perspective is None):
        raise ValueError(
            "represented_party_id 与 represented_party_perspective 必须同时为空或同时有值"
        )


class RepresentingOrganizationItem(BaseModel):
    """代表指定主体（Party）的组织只读摘要（主体详情反向列表用）"""

    organization_id: str = Field(..., description="组织ID")
    name: str = Field(..., description="组织名称")
    code: str = Field(..., description="组织编码")
    level: int = Field(..., description="组织层级")
    status: str = Field(..., description="组织状态")
    parent_id: str | None = Field(None, description="上级组织ID")
    represented_party_perspective: RepresentedPartyPerspective | None = Field(
        None, description="代表视角（owner/manager）"
    )


class OrganizationBase(BaseModel):
    """组织架构基础模式"""

    name: str = Field(..., min_length=1, max_length=200, description="组织名称")
    code: str = Field(..., min_length=1, max_length=50, description="组织编码")
    level: int = Field(default=1, ge=1, le=10, description="组织层级")
    sort_order: int = Field(default=0, ge=0, description="排序")
    parent_id: str | None = Field(None, description="上级组织ID")

    # 组织基本信息
    type: str = Field(..., description="组织类型")
    status: str = Field(..., description="状态")

    # 其他信息
    description: str | None = Field(None, max_length=1000, description="组织描述")


class OrganizationCreate(OrganizationBase):
    """创建组织架构模式"""

    created_by: str | None = Field(None, max_length=100, description="创建人")

    model_config = ConfigDict(extra="forbid")


class OrganizationUpdate(BaseModel):
    """更新组织架构模式"""

    name: str | None = Field(None, min_length=1, max_length=200, description="组织名称")
    code: str | None = Field(None, min_length=1, max_length=50, description="组织编码")
    level: int | None = Field(None, ge=1, le=10, description="组织层级")
    sort_order: int | None = Field(None, ge=0, description="排序")
    # 组织基本信息
    type: str | None = Field(None, description="组织类型")
    status: str | None = Field(None, description="状态")

    # 其他信息
    description: str | None = Field(None, max_length=1000, description="组织描述")

    updated_by: str | None = Field(None, max_length=100, description="更新人")

    model_config = ConfigDict(extra="forbid")


class OrganizationResponse(OrganizationBase):
    """组织架构响应模式"""

    id: str
    path: str | None = None
    is_deleted: bool = False
    created_at: datetime
    updated_at: datetime
    created_by: str | None = None
    updated_by: str | None = None
    represented_party_id: str | None = None
    represented_party_perspective: RepresentedPartyPerspective | None = None

    # 子组织列表
    children: list["OrganizationResponse"] | None = []

    @model_validator(mode="before")
    @classmethod
    def coerce_organization_model(cls, v: Any) -> Any:
        """避免在响应序列化中触发 children 懒加载。"""
        if isinstance(v, dict):
            return v
        try:
            state = sa_inspect(v)
        except Exception:
            return v

        data = {attr.key: getattr(v, attr.key) for attr in state.mapper.column_attrs}

        try:
            children_state = state.attrs.children
            children_value = children_state.loaded_value
            no_value = getattr(orm_attributes, "NO_VALUE", None)
            data["children"] = [] if children_value is no_value else children_value
        except Exception:
            data["children"] = []

        return data

    model_config = ConfigDict(from_attributes=True)


class OrganizationPartyScopeProposal(BaseModel):
    """The only write proposal for an Organization's represented Party."""

    represented_party_id: str | None = Field(
        ..., description="Approved active legal Party ID, or null to inherit"
    )
    represented_party_perspective: RepresentedPartyPerspective | None = Field(
        ..., description="owner or manager, or null to inherit"
    )

    @model_validator(mode="after")
    def validate_represented_party_pair(self) -> "OrganizationPartyScopeProposal":
        _validate_represented_party_pair(
            self.represented_party_id,
            self.represented_party_perspective,
        )
        return self

    model_config = ConfigDict(extra="forbid")


class OrganizationPartyScopeState(BaseModel):
    """Direct link and effective inherited scope for one Organization."""

    represented_party_id: str | None = None
    represented_party_perspective: RepresentedPartyPerspective | None = None
    effective_party_id: str | None = None
    effective_party_perspective: RepresentedPartyPerspective | None = None
    source_organization_id: str | None = None


class OrganizationPartyScopeImpact(BaseModel):
    """Counts affected by an Organization represented-Party change."""

    organization_count: int = Field(..., ge=0)
    organization_scope_change_count: int = Field(..., ge=0)
    user_count: int = Field(..., ge=0)
    user_scope_change_count: int = Field(..., ge=0)


class OrganizationPartyScopePreviewResponse(BaseModel):
    organization_id: str
    before_scope: OrganizationPartyScopeState
    after_scope: OrganizationPartyScopeState
    impact: OrganizationPartyScopeImpact
    preview_token: str
    expires_at: datetime


class OrganizationPartyScopeCommitRequest(BaseModel):
    preview_token: str = Field(..., min_length=1, max_length=512)
    reason: str = Field(..., min_length=1, max_length=500)
    idempotency_key: str = Field(..., min_length=1, max_length=128)

    @field_validator("preview_token", "reason", "idempotency_key")
    @classmethod
    def reject_blank_value(cls, value: str) -> str:
        normalized = value.strip()
        if normalized == "":
            raise ValueError("value must not be blank")
        return normalized

    model_config = ConfigDict(extra="forbid")


class OrganizationPartyScopeCommitResponse(BaseModel):
    organization: OrganizationResponse
    before_scope: OrganizationPartyScopeState
    after_scope: OrganizationPartyScopeState
    impact: OrganizationPartyScopeImpact
    committed_at: datetime
    idempotent: bool = False


class OrganizationPartyScopeBatchProposal(OrganizationPartyScopeProposal):
    """One direct Organization Party scope proposal inside a batch."""

    organization_id: str = Field(..., min_length=1, max_length=100)

    @field_validator("organization_id")
    @classmethod
    def normalize_organization_id(cls, value: str) -> str:
        normalized = value.strip()
        if normalized == "":
            raise ValueError("organization_id must not be blank")
        return normalized


class OrganizationPartyScopeBatchPreviewRequest(BaseModel):
    items: list[OrganizationPartyScopeBatchProposal] = Field(
        ..., min_length=2, max_length=100
    )

    @model_validator(mode="after")
    def reject_duplicate_organizations(
        self,
    ) -> "OrganizationPartyScopeBatchPreviewRequest":
        organization_ids = [item.organization_id for item in self.items]
        if len(set(organization_ids)) != len(organization_ids):
            raise ValueError("organization_id must be unique within a batch")
        return self

    model_config = ConfigDict(extra="forbid")


class OrganizationPartyScopeBatchPreviewItem(BaseModel):
    organization: OrganizationResponse
    before_scope: OrganizationPartyScopeState
    after_scope: OrganizationPartyScopeState
    impact: OrganizationPartyScopeImpact


class OrganizationPartyScopeBatchPreviewResponse(BaseModel):
    items: list[OrganizationPartyScopeBatchPreviewItem]
    impact: OrganizationPartyScopeImpact
    preview_token: str
    expires_at: datetime


class OrganizationPartyScopeBatchCommitRequest(OrganizationPartyScopeCommitRequest):
    """Confirmation request for a previously generated batch preview."""


class OrganizationPartyScopeBatchCommitResponse(BaseModel):
    items: list[OrganizationPartyScopeBatchPreviewItem]
    impact: OrganizationPartyScopeImpact
    committed_at: datetime
    idempotent: bool = False


class OrganizationMoveProposal(BaseModel):
    """The only write proposal for an Organization hierarchy move."""

    target_parent_id: str | None = Field(
        ..., description="Target parent Organization ID, or null to become a root"
    )

    @field_validator("target_parent_id")
    @classmethod
    def normalize_target_parent_id(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        if normalized == "":
            raise ValueError("target_parent_id must not be blank")
        return normalized

    model_config = ConfigDict(extra="forbid")


class OrganizationMoveScopeState(BaseModel):
    """Effective Party scope at the root of a proposed Organization move."""

    parent_id: str | None = None
    effective_party_id: str | None = None
    effective_party_perspective: RepresentedPartyPerspective | None = None
    source_organization_id: str | None = None


class OrganizationMoveImpact(BaseModel):
    """Counts affected by an Organization hierarchy move."""

    organization_count: int = Field(..., ge=0)
    organization_scope_change_count: int = Field(..., ge=0)
    organization_path_change_count: int = Field(..., ge=0)
    user_count: int = Field(..., ge=0)
    user_scope_change_count: int = Field(..., ge=0)


class OrganizationMovePreviewResponse(BaseModel):
    organization_id: str
    before_scope: OrganizationMoveScopeState
    after_scope: OrganizationMoveScopeState
    impact: OrganizationMoveImpact
    preview_token: str
    expires_at: datetime


class OrganizationMoveCommitRequest(BaseModel):
    preview_token: str = Field(..., min_length=1, max_length=512)
    reason: str = Field(..., min_length=1, max_length=500)
    idempotency_key: str = Field(..., min_length=1, max_length=128)

    @field_validator("preview_token", "reason", "idempotency_key")
    @classmethod
    def reject_blank_value(cls, value: str) -> str:
        normalized = value.strip()
        if normalized == "":
            raise ValueError("value must not be blank")
        return normalized

    model_config = ConfigDict(extra="forbid")


class OrganizationMoveCommitResponse(BaseModel):
    organization: OrganizationResponse
    before_scope: OrganizationMoveScopeState
    after_scope: OrganizationMoveScopeState
    impact: OrganizationMoveImpact
    committed_at: datetime
    idempotent: bool = False


class OrganizationTree(BaseModel):
    """组织架构树形结构"""

    id: str
    name: str
    code: str
    level: int
    sort_order: int
    type: str
    status: str
    children: list["OrganizationTree"] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class OrganizationHistoryResponse(BaseModel):
    """组织架构历史响应模式"""

    id: str
    organization_id: str
    action: str
    field_name: str | None = None
    old_value: str | None = None
    new_value: str | None = None
    change_reason: str | None = None
    created_at: datetime
    created_by: str | None = None

    model_config = ConfigDict(from_attributes=True)


class OrganizationStatistics(BaseModel):
    """组织架构统计信息"""

    total: int = Field(..., description="总数")
    active: int = Field(..., description="活跃数量")
    inactive: int = Field(..., description="非活跃数量")
    by_type: dict[str, Any] = Field(..., description="按类型统计")
    by_level: dict[str, Any] = Field(..., description="按层级统计")


class OrganizationSearchRequest(BaseModel):
    """组织搜索请求"""

    keyword: str | None = Field(None, description="关键词")
    level: int | None = Field(None, ge=1, le=10, description="层级")
    parent_id: str | None = Field(None, description="上级组织ID")
    skip: int = Field(0, ge=0, description="跳过数量")
    page_size: int = Field(100, ge=1, le=1000, description="每页大小")


# 更新前向引用
OrganizationResponse.model_rebuild()
OrganizationTree.model_rebuild()

File: frontend/src/types/organization.ts
/**
 * 组织架构相关类型定义
 * @deprecated Phase 3 兼容保留，待 Phase 4 完成 Party 全量迁移后移除。
 */

export interface Organization {
  id: string;
  name: string;
  code: string;
  level: number;
  sort_order: number;
  parent_id?: string;
  path?: string;

  // 组织基本信息
  type: string;
  status: string;
  represented_party_id?: string | null;
  represented_party_perspective?: OrganizationPartyPerspective | null;

  // 其他信息
  description?: string;

  // 系统字段
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  created_by?: string;
  updated_by?: string;

  children?: Organization[];
}

export interface OrganizationCreate {
  name: string;
  code: string;
  level?: number;
  sort_order?: number;
  parent_id?: string;
  type: string;
  status: string;

  // 其他信息
  description?: string;

  created_by?: string;
}

export interface OrganizationUpdate {
  name?: string;
  code?: string;
  level?: number;
  sort_order?: number;
  parent_id?: string;
  type?: string;
  status?: string;

  // 其他信息
  description?: string;

  updated_by?: string;
}

export type OrganizationPartyPerspective = 'owner' | 'manager';

export interface OrganizationPartyScopeProposal {
  represented_party_id: string | null;
  represented_party_perspective: OrganizationPartyPerspective | null;
}

export interface OrganizationPartyScopeState {
  represented_party_id?: string | null;
  represented_party_perspective?: OrganizationPartyPerspective | null;
  effective_party_id?: string | null;
  effective_party_perspective?: OrganizationPartyPerspective | null;
  source_organization_id?: string | null;
}

export interface OrganizationPartyScopeImpact {
  organization_count: number;
  organization_scope_change_count: number;
  user_count: number;
  user_scope_change_count: number;
}

export interface OrganizationPartyScopePreview {
  organization_id: string;
  before_scope: OrganizationPartyScopeState;
  after_scope: OrganizationPartyScopeState;
  impact: OrganizationPartyScopeImpact;
  preview_token: string;
  expires_at: string;
}

export interface OrganizationPartyScopeCommitRequest {
  preview_token: string;
  reason: string;
  idempotency_key: string;
}

export interface OrganizationPartyScopeCommitResponse {
  organization: Organization;
  before_scope: OrganizationPartyScopeState;
  after_scope: OrganizationPartyScopeState;
  impact: OrganizationPartyScopeImpact;
  committed_at: string;
  idempotent: boolean;
}

export interface OrganizationPartyScopeBatchProposal extends OrganizationPartyScopeProposal {
  organization_id: string;
}

export interface OrganizationPartyScopeBatchPreviewItem {
  organization: Organization;
  before_scope: OrganizationPartyScopeState;
  after_scope: OrganizationPartyScopeState;
  impact: OrganizationPartyScopeImpact;
}

export interface OrganizationPartyScopeBatchPreviewRequest {
  items: OrganizationPartyScopeBatchProposal[];
}

export interface OrganizationPartyScopeBatchPreview {
  items: OrganizationPartyScopeBatchPreviewItem[];
  impact: OrganizationPartyScopeImpact;
  preview_token: string;
  expires_at: string;
}

export interface OrganizationPartyScopeBatchCommitRequest {
  preview_token: string;
  reason: string;
  idempotency_key: string;
}

export interface OrganizationPartyScopeBatchCommitResponse {
  items: OrganizationPartyScopeBatchPreviewItem[];
  impact: OrganizationPartyScopeImpact;
  committed_at: string;
  idempotent: boolean;
}

export interface OrganizationTree {
  id: string;
  name: string;
  code: string;
  level: number;
  sort_order: number;
  type: string;
  status: string;
  children: OrganizationTree[];
}

export interface OrganizationHistory {
  id: string;
  organization_id: string;
  action: 'create' | 'update' | 'delete' | 'move' | 'party_scope_update';
  field_name?: string;
  old_value?: string;
  new_value?: string;
  change_reason?: string;
  created_at: string;
  created_by?: string;
}

export interface OrganizationStatistics {
  total: number;
  active: number;
  inactive: number;
  by_level: Record<string, number>;
  by_type: Record<string, number>;
}

export interface OrganizationMoveProposal {
  target_parent_id: string | null;
}

export interface OrganizationMoveScopeState {
  parent_id: string | null;
  effective_party_id: string | null;
  effective_party_perspective: 'owner' | 'manager' | null;
  source_organization_id: string | null;
}

export interface OrganizationMoveImpact {
  organization_count: number;
  organization_scope_change_count: number;
  organization_path_change_count: number;
  user_count: number;
  user_scope_change_count: number;
}

export interface OrganizationMovePreview {
  organization_id: string;
  before_scope: OrganizationMoveScopeState;
  after_scope: OrganizationMoveScopeState;
  impact: OrganizationMoveImpact;
  preview_token: string;
  expires_at: string;
}

export interface OrganizationMoveCommitRequest {
  preview_token: string;
  reason: string;
  idempotency_key: string;
}

export interface OrganizationMoveCommitResponse {
  organization: Organization;
  before_scope: OrganizationMoveScopeState;
  after_scope: OrganizationMoveScopeState;
  impact: OrganizationMoveImpact;
  committed_at: string;
  idempotent: boolean;
}

export interface OrganizationSearchRequest {
  keyword?: string;
  level?: number;
  parent_id?: string;
  page?: number;
  page_size?: number;
}

// API响应类型
export interface OrganizationListResponse {
  data: Organization[];
  total: number;
  page: number;
  page_size: number;
}

export interface OrganizationResponse {
  data: Organization;
  message?: string;
}

export interface OrganizationTreeResponse {
  data: OrganizationTree[];
}

export interface OrganizationStatisticsResponse {
  data: OrganizationStatistics;
}

export interface OrganizationHistoryResponse {
  data: OrganizationHistory[];
  total: number;
  page: number;
  page_size: number;
}

// 表单验证规则
export interface OrganizationFormRules {
  name: Array<{ required: boolean; message: string; min?: number; max?: number }>;
}

// 树形选择器数据节点
export interface OrganizationTreeNode {
  key: string;
  title: React.ReactNode;
  value: string;
  children?: OrganizationTreeNode[];
  disabled?: boolean;
}

// 组织路径信息
export interface OrganizationPath {
  organizations: Organization[];
  path_string: string;
}

// 高级搜索条件
export interface OrganizationSearchCriteria extends OrganizationSearchRequest {
  date_range?: [string, string];
  created_by?: string;
  updated_by?: string;
}

// 组织导出配置
export interface OrganizationExportConfig {
  format: 'excel' | 'csv' | 'pdf';
  fields: string[];
  include_children: boolean;
  filter?: OrganizationSearchRequest;
}

// 组织导入配置
export interface OrganizationImportConfig {
  file: File;
  update_existing: boolean;
  skip_errors: boolean;
  dry_run: boolean;
}

// 组织导入结果
export interface OrganizationImportResult {
  success_count: number;
  error_count: number;
  errors: Array<{
    row: number;
    field: string;
    message: string;
  }>;
  preview?: Organization[];
}
